File: Decorato.java/src/com/parttern/decorator/BlackSugarBubble.java
package com.parttern.decorator;

public class BlackSugarBubble extends ToppingDecorator {
    private double price = 0.2;

    public BlackSugarBubble(MilkTea innerMilkTea) {
        super(innerMilkTea);
    }

    @Override
    public double cost() {
        return this.price + milkTea.cost();
    }

    @Override
    public String doMilkTea() {
        String type = milkTea.doMilkTea();
        return type + addBlackSugarBubble();
    }

    private String addBlackSugarBubble() {
        return " ,Trân châu đường đen";
    }
}

File: Decorator/BlackSugarBubble.cs
namespace Decorator {
    internal class BlackSugarBubble : ToppingDecorator {
        private readonly double price = 0.2;

        public BlackSugarBubble(MilkTea milkTea) : base(milkTea) {
        }

        public override string DoMilkTea() {
            return milkTea.DoMilkTea() + AddBlackSugarBubble();
        }

        private string AddBlackSugarBubble() {
            return ", Trân châu đường đen";
        }

        public override double Cost() {
            return price + milkTea.Cost();
        }
    }
}
